src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs

```diff
@@ -0,0 +1,1845 @@
+            yield return getData("static () => 1", "System.Func<System.Int32>");
+
+            static object?[] getData(string expr, string? expectedType) =>
+                new object?[] { expr, expectedType };
+        }
+
+        [Theory]
+        [MemberData(nameof(GetExpressionData))]
+        public void Expression_ImplicitConversion(string anonymousFunction, string? expectedType)
+        {
+            var source =
+$@"class Program
+{{
+    static void Main(string[] args)
+    {{
+        System.Linq.Expressions.Expression e = {anonymousFunction};
+    }}
+}}";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.RegularPreview);
+            if (expectedType is null)
+            {
+                comp.VerifyDiagnostics(
+                    // (5,48): error CS8915: The delegate type could not be inferred.
+                    //         System.Linq.Expressions.Expression e = x => x;
+                    Diagnostic(ErrorCode.ERR_CannotInferDelegateType, anonymousFunction).WithLocation(5, 48));
+            }
+            else
+            {
+                comp.VerifyDiagnostics();
+            }
+
+            var tree = comp.SyntaxTrees[0];
+            var model = comp.GetSemanticModel(tree);
+            var expr = tree.GetRoot().DescendantNodes().OfType<AnonymousFunctionExpressionSyntax>().Single();
+            var typeInfo = model.GetTypeInfo(expr);
+            if (expectedType == null)
+            {
+                Assert.Null(typeInfo.Type);
+            }
+            else
+            {
+                Assert.Equal($"System.Linq.Expressions.Expression<{expectedType}>", typeInfo.Type.ToTestDisplayString());
+            }
+            Assert.Equal("System.Linq.Expressions.Expression", typeInfo.ConvertedType!.ToTestDisplayString());
+        }
+
+        [Theory]
+        [MemberData(nameof(GetExpressionData))]
+        public void Expression_ExplicitConversion(string anonymousFunction, string? expectedType)
+        {
+            var source =
+$@"class Program
+{{
+    static void Main(string[] args)
+    {{
+        object o = (System.Linq.Expressions.Expression)({anonymousFunction});
+    }}
+}}";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.RegularPreview);
+            if (expectedType is null)
+            {
+                comp.VerifyDiagnostics(
+                    // (5,20): error CS8915: The delegate type could not be inferred.
+                    //         object o = (System.Linq.Expressions.Expression)(x => x);
+                    Diagnostic(ErrorCode.ERR_CannotInferDelegateType, $"(System.Linq.Expressions.Expression)({anonymousFunction})").WithLocation(5, 20));
+            }
+            else
+            {
+                comp.VerifyDiagnostics();
+            }
```

Comment:
Consider asserting TypeInfo here as well.

Reply:
Same here.

---
In reply to: [618754496](https://github.com/dotnet/roslyn/pull/52448#discussion_r618754496) [](ancestors = 618754496)